hi-ml-histopathology/src/histopathology/utils/report_utils.py

```diff
@@ -0,0 +1,339 @@
+#  -------------------------------------------------------------------------------------------
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  -------------------------------------------------------------------------------------------
+
+import pickle
+from pathlib import Path
+from typing import Dict, Optional, Sequence
+
+import dateutil.parser
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from azureml.core import Experiment, Run, Workspace
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+from matplotlib.lines import Line2D
+from sklearn.metrics import auc, precision_recall_curve, roc_curve
+
+from health_azure.utils import aggregate_hyperdrive_metrics, get_aml_run_from_run_id, get_tags_from_hyperdrive_run
+from histopathology.utils.naming import ResultsKey
+
+TRAIN_STYLE = dict(ls='-')
+VAL_STYLE = dict(ls='--')
+BEST_EPOCH_LINE_STYLE = dict(ls=':', lw=1)
+BEST_TRAIN_MARKER_STYLE = dict(marker='o', markeredgecolor='w', markersize=6)
+BEST_VAL_MARKER_STYLE = dict(marker='*', markeredgecolor='w', markersize=11)
+
+
+def download_from_run_if_necessary(run: Run, remote_dir: Path, download_dir: Path, filename: str) -> Path:
+    """Download any file from an AML run if it doesn't exist locally.
+
+    :param run: AML Run object.
+    :param remote_dir: Remote directory from where the file is downloaded.
+    :param download_dir: Local directory where to save the downloaded file.
+    :param filename: Name of the file to be downloaded (e.g. `"test_output.csv"`).
+    :return: Local path to the downloaded file.
+    """
+    local_path = download_dir / filename
+    remote_path = remote_dir / filename
+    if local_path.exists():
+        print("File already exists at", local_path)
+    else:
+        local_path.parent.mkdir(exist_ok=True, parents=True)
+        run.download_file(str(remote_path), str(local_path), _validate_checksum=True)
+        assert local_path.exists()
+        print("File is downloaded at", local_path)
+    return local_path
+
+
+def collect_crossval_outputs(parent_run_id: str, download_dir: Path, aml_workspace: Workspace,
```

Comment:
There is a function in health_azure already called `download_files_from_hyperdrive_children` which will download these files and return a list of download paths. You could call it in this function (not sure if it is necessary to avoid overwriting, especially since there is nothing here checking the run is complete, so you could end up downloading incomplete metrics files while the run is still ongoing?)

Reply:
Maybe we can make it optional to skip over existing files? Or at least add a warning to the user that files will be skipped?

Reply:
As discussed, added `overwrite` flag and migrated `download_file_if_necessary()` to `health_azure.utils`.